File: TRENDX/Repositories/AuthRepository.swift
//
//  AuthRepository.swift
//  TRENDX
//

import Foundation
#if canImport(UIKit)
import UIKit
#endif

struct AuthSession: Codable, Equatable {
    var accessToken: String
    var refreshToken: String?
    var userId: UUID
    var email: String
}

final class AuthRepository {
    private let client: TrendXAPIClient
    private let defaults: UserDefaults
    private let sessionKey = "trendx_auth_session_v1"

    init(client: TrendXAPIClient, defaults: UserDefaults = .standard) {
        self.client = client
        self.defaults = defaults
    }

    var isRemoteEnabled: Bool {
        client.config.isConfigured
    }

    func restoreSession() -> AuthSession? {
        guard let data = defaults.data(forKey: sessionKey) else { return nil }
        return try? JSONDecoder().decode(AuthSession.self, from: data)
    }

    func signIn(email: String, password: String) async throws -> AuthSession {
        if !client.config.isConfigured {
            let session = localSession(email: email)
            save(session)
            return session
        }

        let response: AuthResponse = try await client.post(
            "/auth/signin",
            body: AuthCredentials(email: email, password: password)
        )
        let session = response.session
        save(session)
        return session
    }

    func signUp(
        name: String,
        email: String,
        password: String,
        gender: UserGender = .unspecified,
        birthYear: Int? = nil,
        city: String? = nil,
        region: String? = nil
    ) async throws -> AuthSession {
        if !client.config.isConfigured {
            let session = localSession(email: email)
            save(session)
            return session
        }

        let payload = SignUpPayload(
            name: name,
            email: email,
            password: password,
            gender: gender.rawValue,
            birthYear: birthYear,
            city: city,
            region: region,
            deviceType: "ios",
            osVersion: Self.systemVersionString
        )
        let response: AuthResponse = try await client.post(
            "/auth/signup",
            body: payload
        )

        let session = response.session
        save(session)
        return session
    }

    func signOut() {
        defaults.removeObject(forKey: sessionKey)
    }

    func fetchProfile(session: AuthSession) async throws -> TrendXUser {
        let profile: UserDTO = try await client.get("/profile", accessToken: session.accessToken)
        return profile.domain
    }

    func upsertProfile(name: String, email: String, session: AuthSession) async throws {
        let payload = ProfilePayload(
            id: session.userId,
            name: name,
            email: email,
            avatarInitial: String(name.prefix(1)),
            points: 100,
            coins: 16.67,
            isPremium: false
        )
        let _: UserDTO = try await client.post("/profile", accessToken: session.accessToken, body: payload)
    }

    /// Update arbitrary profile fields. Only non-nil keys are sent to the
    /// backend, so callers can selectively patch (name + city) without
    /// nuking the rest. Returns the freshly-decoded domain user.
    func updateProfile(
        name: String? = nil,
        email: String? = nil,
        handle: String? = nil,
        bio: String? = nil,
        avatarInitial: String? = nil,
        avatarUrl: String? = nil,
        bannerUrl: String? = nil,
        accountType: AccountType? = nil,
        gender: String? = nil,
        birthYear: Int? = nil,
        city: String? = nil,
        region: String? = nil,
        country: String? = nil,
        session: AuthSession
    ) async throws -> TrendXUser {
        let payload = ProfileUpdatePayload(
            name: name,
            email: email,
            handle: handle,
            bio: bio,
            avatarInitial: avatarInitial,
            avatarUrl: avatarUrl,
            bannerUrl: bannerUrl,
            accountType: accountType?.rawValue,
            gender: gender,
            birthYear: birthYear,
            city: city,
            region: region,
            country: country
        )
        let dto: UserDTO = try await client.post("/profile", accessToken: session.accessToken, body: payload)
        return dto.domain
    }

    /// Cheap server-side availability check for an `@handle` candidate.
    /// Returns nil when the handle is OK, or a localized error message
    /// when it's invalid / reserved / taken.
    func checkHandleAvailability(_ candidate: String, session: AuthSession) async throws -> String? {
        struct Response: Decodable {
            let ok: Bool
            let reason: String?
            let message: String?
        }
        let escaped = candidate.addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? candidate
        let result: Response = try await client.get(
            "/handles/check?handle=\(escaped)",
            accessToken: session.accessToken
        )
        return result.ok ? nil : (result.message ?? "هذا المعرّف غير متاح.")
    }

    private func save(_ session: AuthSession) {
        if let data = try? JSONEncoder().encode(session) {
            defaults.set(data, forKey: sessionKey)
        }
    }

    private func localSession(email: String) -> AuthSession {
        AuthSession(
            accessToken: "local-beta-token",
            refreshToken: nil,
            userId: UUID(uuidString: "00000000-0000-0000-0000-000000000001") ?? UUID(),
            email: email
        )
    }

    private static var systemVersionString: String {
        #if canImport(UIKit)
        return UIDevice.current.systemVersion
        #else
        let v = ProcessInfo.processInfo.operatingSystemVersion
        return "\(v.majorVersion).\(v.minorVersion).\(v.patchVersion)"
        #endif
    }
}

private struct ProfileUpdatePayload: Encodable {
    let name: String?
    let email: String?
    let handle: String?
    let bio: String?
    let avatarInitial: String?
    let avatarUrl: String?
    let bannerUrl: String?
    let accountType: String?
    let gender: String?
    let birthYear: Int?
    let city: String?
    let region: String?
    let country: String?

    func encode(to encoder: Encoder) throws {
        var container = encoder.container(keyedBy: CodingKeys.self)
        if let name { try container.encode(name, forKey: .name) }
        if let email { try container.encode(email, forKey: .email) }
        if let handle { try container.encode(handle, forKey: .handle) }
        if let bio { try container.encode(bio, forKey: .bio) }
        if let avatarInitial { try container.encode(avatarInitial, forKey: .avatarInitial) }
        if let avatarUrl { try container.encode(avatarUrl, forKey: .avatarUrl) }
        if let bannerUrl { try container.encode(bannerUrl, forKey: .bannerUrl) }
        if let accountType { try container.encode(accountType, forKey: .accountType) }
        if let gender { try container.encode(gender, forKey: .gender) }
        if let birthYear { try container.encode(birthYear, forKey: .birthYear) }
        if let city { try container.encode(city, forKey: .city) }
        if let region { try container.encode(region, forKey: .region) }
        if let country { try container.encode(country, forKey: .country) }
    }

    enum CodingKeys: String, CodingKey {
        case name, email, handle, bio,
             avatarInitial, avatarUrl, bannerUrl,
             accountType, gender, birthYear, city, region, country
    }
}

private struct AuthCredentials: Encodable {
    let email: String
    let password: String
}

private struct SignUpPayload: Encodable {
    let name: String
    let email: String
    let password: String
    let gender: String
    let birthYear: Int?
    let city: String?
    let region: String?
    let deviceType: String
    let osVersion: String?
}

private struct AuthResponse: Decodable {
    let accessToken: String?
    let refreshToken: String?
    let user: AuthUser

    var session: AuthSession {
        AuthSession(
            accessToken: accessToken ?? "",
            refreshToken: refreshToken,
            userId: user.id,
            email: user.email ?? ""
        )
    }
}

private struct AuthUser: Decodable {
    let id: UUID
    let email: String?
}

private struct ProfilePayload: Codable {
    let id: UUID
    let name: String
    let email: String
    let avatarInitial: String
    let points: Int
    let coins: Double
    let isPremium: Bool
}

File: android/app/src/main/java/com/trendx/app/repositories/AuthRepository.kt
package com.trendx.app.repositories

import com.trendx.app.models.TrendXUser
import com.trendx.app.models.UserGender
import com.trendx.app.networking.AuthCredentials
import com.trendx.app.networking.AuthResponse
import com.trendx.app.networking.HandleCheckResponse
import com.trendx.app.networking.SignUpPayload
import com.trendx.app.networking.TrendXAPIClient
import com.trendx.app.networking.UserDto
import io.ktor.http.encodeURLParameter

// Mirrors TRENDX/Repositories/AuthRepository.swift. When the backend isn't
// configured (rare on Android — BuildConfig sets it at build time), we
// synthesize a local session so screens can still render with samples.
class AuthRepository(
    private val client: TrendXAPIClient,
    private val sessionStore: SessionStore
) {
    val isRemoteEnabled: Boolean get() = client.config.isConfigured

    suspend fun restoreSession(): AuthSession? = sessionStore.currentSession()

    suspend fun signIn(email: String, password: String): AuthSession {
        if (!isRemoteEnabled) {
            return localSession(email).also { sessionStore.save(it) }
        }
        val response: AuthResponse = client.post(
            "/auth/signin",
            body = AuthCredentials(email, password)
        )
        return response.toSession().also { sessionStore.save(it) }
    }

    suspend fun signUp(
        name: String,
        email: String,
        password: String,
        gender: UserGender = UserGender.unspecified,
        birthYear: Int? = null,
        city: String? = null,
        region: String? = null,
        osVersion: String? = null
    ): AuthSession {
        if (!isRemoteEnabled) {
            return localSession(email).also { sessionStore.save(it) }
        }
        val response: AuthResponse = client.post(
            "/auth/signup",
            body = SignUpPayload(
                name = name,
                email = email,
                password = password,
                gender = gender.name,
                birthYear = birthYear,
                city = city,
                region = region,
                osVersion = osVersion
            )
        )
        return response.toSession().also { sessionStore.save(it) }
    }

    suspend fun signOut() = sessionStore.clear()

    suspend fun fetchProfile(session: AuthSession): TrendXUser {
        val dto: UserDto = client.get("/profile", accessToken = session.accessToken)
        return dto.toDomain()
    }

    /// POST `/profile` with the editable subset of fields. Backend reads
    /// snake_case (the Json naming strategy converts automatically). All
    /// fields are optional — only the ones the user actually changed are
    /// sent, blank strings are normalized to empty strings (the backend
    /// treats them the same as null on PATCH-style updates).
    suspend fun updateProfile(
        name: String? = null,
        email: String? = null,
        handle: String? = null,
        bio: String? = null,
        avatarUrl: String? = null,
        bannerUrl: String? = null,
        accountType: String? = null,
        gender: String? = null,
        birthYear: Int? = null,
        city: String? = null,
        country: String? = null,
        session: AuthSession
    ): TrendXUser {
        val payload = ProfileUpdatePayload(
            name = name,
            email = email,
            handle = handle,
            bio = bio,
            avatarUrl = avatarUrl,
            bannerUrl = bannerUrl,
            accountType = accountType,
            gender = gender,
            birthYear = birthYear,
            city = city,
            country = country
        )
        val dto: UserDto = client.post(
            path = "/profile",
            body = payload,
            accessToken = session.accessToken
        )
        return dto.toDomain()
    }

    suspend fun checkHandleAvailability(candidate: String, session: AuthSession): String? {
        val escaped = candidate.encodeURLParameter()
        val response: HandleCheckResponse = client.get(
            "/handles/check?handle=$escaped",
            accessToken = session.accessToken
        )
        return if (response.ok) null else (response.message ?: "هذا المعرّف غير متاح.")
    }

    private fun AuthResponse.toSession() = AuthSession(
        accessToken = accessToken.orEmpty(),
        refreshToken = refreshToken,
        userId = user.id,
        email = user.email.orEmpty()
    )

    private fun localSession(email: String) = AuthSession(
        accessToken = "local-beta-token",
        refreshToken = null,
        userId = "00000000-0000-0000-0000-000000000001",
        email = email
    )
}

@kotlinx.serialization.Serializable
private data class ProfileUpdatePayload(
    val name: String? = null,
    val email: String? = null,
    val handle: String? = null,
    val bio: String? = null,
    val avatarUrl: String? = null,
    val bannerUrl: String? = null,
    val accountType: String? = null,
    val gender: String? = null,
    val birthYear: Int? = null,
    val city: String? = null,
    val country: String? = null
)
